### DISCUSSION TITLE
Use vue-facing-decorator in a browser only, with no build step

### DISCUSSION THREAD

**Discussion by johandanforth** [facing-dev/vue-facing-decorator#40] [Q&A]
I'm looking at using Vue 3 and vue-facing-decorator in an environment that doesn't have Internet access, and without a build-step. I'm using referencing "vue" through the help of an importmap like this:

``` html
<script type="importmap">
	{
	  "imports": {
		"vue": "https://unpkg.com/vue@3/dist/vue.esm-browser.js",
	  }
	}
</script>

<div id="app">
	<div>{{ message }}</div>
</div>

<script type="module">
	import { createApp } from 'vue'

	createApp({
		data() {
			return {
				message: 'Hello Vue!'
			}
		}
	}).mount('#app')
</script>

```

Is it possible to reference vue-facing-decorator in the same way, through importmap and browser-only? The package contains many different files, I'm not sure where to start :D 

Thanks for any help!

**Comment by ruojianll** [comment]
v-f-d in npm is a esm build. You could try import it's index.js file in importmap script. If that not works, I could build a single esm file version for you.

**Comment by johandanforth** [comment]
I've tried various ways of importing like you suggest, but there are too many reference to packages in an npm style, that it's impossible. I've tried with both importmaps and path alias. 

I would really appreciate it if you could build a "browser version"! 

I'm developing in an environment with no internet or npm access, so it's a bit of a hassle 🤪

**Comment by ruojianll** [comment]
I released v2.1.14, that contains a esm bulid in `dist/esm/xxx`. The code works:

```
    <script type="importmap">
        {
            "imports":{
                "vue": "https://unpkg.com/vue@3/dist/vue.esm-browser.js",
                "vfd":"https://unpkg.com/vue-facing-decorator@2.1.14/dist/esm/index.js"
            }
        }
    </script>
    <script type="module">
        import * as VFD from 'vfd'
        console.log('v-f-d',VFD)
    </script>
```

**Reply by johandanforth** [reply]
Wow, that was quick! I'll try it out today and post back how it went 🙏

**Reply by johandanforth** [reply]
Importing works fine in script section in the brower, but I'm not able to import from a TypeScript file. That gives me loads and loads of import errors for the underlying bits. I'm sorry for being totally lost, I'm still new to the import and export stuff and the jungle of various styles of modules.

What I would like to do is having both Vue 3 and your library inside my own local project, all files, and be able to define both apps and components from TypeScript files.

**Reply by ruojianll** [reply]
If you want to use typescript, you should include a build step. Typescript is a compiler not a runtime.

**Reply by johandanforth** [reply]
I'm well aware of how TypeScript is and can be used. I'm transpiling TS at compile time of my dotnet project. But I'm not using webpack or node. It's perfectly fine to use TS like this, except that imports is a bit of a hassle 🤪

But I'm probably going to move over to use Vue CLI, even though it involves downloading a gazillion of random js libraries...

**Reply by ruojianll** [reply]
In my opinion, your could try install/import v-f-d in your local npm ts project and `tsc` your project in `esm` module. You could use v-f-d's declaration files and it won't be compiled in the final js files. Then just import the compiled file in your web file just like above code.

**Reply by johandanforth** [reply]
Thanks for your answers, really! I'm not sure I fully understand though. You're suggesting installing v-f-d with `npm -i` then configure typescript to use ECMAScript Modules `"module": "nodenext"` and specify `"type": "module"` in package.json? Then I should be able to `import...` v-f-d in my TS files? Am I on the right track?

It's a complete jungle with these different module systems for me... :/

**Reply by ruojianll** [reply]
Just set `"module":"ESNext` in `tsconfig.json`. Then develop your app and run `tsc` as usual, the output files could be linked in your html file directly.

main.ts
```typescript
import {Component,Vue} from 'vue-facing-decorator'
@Component
export default class MyComp extends Vue{
  myData:string = 'abc'
} 
```

run `tsc`

output main.js
```javascript
import {Component,Vue} from 'vue-facing-decorator'
@Component
export default class MyComp extends Vue{
  myData = 'abc'
}
```

config importmap in html file
```
  <script type="importmap">
        {
            "imports":{
                "vue": "https://unpkg.com/vue@3/dist/vue.esm-browser.js",
                "vue-facing-decorator":"https://unpkg.com/vue-facing-decorator@2.1.14/dist/esm/index.js",
                "main":"xxx/main.js"
            }
        }
    </script>
<script>
import {MyComp} from 'main'
//init your vue app with MyComp
</script>
```

**Reply by johandanforth** [reply]
Again, I appreciate your efforts to help me, really. I'm making some progress here but not all the way. Follow me...

**EDIT** first of all, I had to install vue and vue-facing-decorators so that I got them in my node_modules, otherwise `tsc` wouldn't compile:

`npm install --save vue`
`npm install --save vue-facing-decorator` 

tsconfig.json:
``` json
{
  "compileOnSave": true,
  "compilerOptions": {
    "noImplicitAny": false,
    "noEmitOnError": true,
    "removeComments": false,
    "sourceMap": true,
    "lib": [
      "es2018",
      "dom"
    ],
    "target": "esnext",
    "module": "esnext",
    "moduleResolution": "node",
    "experimentalDecorators": true
  },
  "include": [
    "wwwroot/ts"
  ],
    "exclude": [
    "node_modules"
  ]
}

```
html/cshtml/razor file:
``` html
<!-- stuff above here -->
<script type="importmap">
	{
		"imports":{
			"vue": "https://unpkg.com/vue@3/dist/vue.esm-browser.js",
			"vue-facing-decorator":"https://unpkg.com/vue-facing-decorator@2.1.14/dist/esm/index.js",
			"MyComp":"/ts/MyComp.js"
		}
	}
</script>
<script type="module">
	import MyComp  from 'MyComp'

	//init vue app with MyComp here...

</script>

```

MyComp.ts:
``` typescript
import { Component, Vue } from 'vue-facing-decorator'
@Component
export default class MyComp extends Vue { 
    myData: string = 'asd'
} 

```

It compiles fine but the js-file is not looking like what you got above, mine is like this.

MyComp.js:
``` javascript
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Component, Vue } from 'vue-facing-decorator';
let MyComp = class MyComp extends Vue {
    myData = 'asd';
};
MyComp = __decorate([
    Component
], MyComp);
export default MyComp;
//# sourceMappingURL=MyComp.js.map

```
And, when I browse the page I get this error from v-f-d:

```
Uncaught TypeError: Cannot read properties of null (reading 'constructor')
    at getSuperSlot (utils.ts:82:17)
    at build (component.ts:87:23)
    at component.ts:118:9
    at _Component (component.ts:103:16)
    at Component (component.ts:117:12)
    at __decorate (MyComp.js:4:95)
    at MyComp.ts:3:28
```

It may have to do with the fact that the Vue constructor needs to be called with `super` and it requires 2 parameters. I'm stuck here.

Any ideas how to progress? I guess you didn't try this yourself? 

Again, appreciate your help, If I can get v-f-d working, it will come to great use for me and I will gladly donate at least 3-4 beers :D

**Reply by ruojianll** [reply]
Thank you for your approval. After delete `.js` of vue and v-f-d urls, the code works.

```
<script type="importmap">
	{
		"imports":{
			"vue": "https://unpkg.com/vue@3/dist/vue.esm-browser",
			"vue-facing-decorator":"https://unpkg.com/vue-facing-decorator@2.1.14/dist/esm/index",
			"MyComp":"/ts/MyComp.js"
		}
	}
</script>
```

Because the referenced file names shoule be same in sourcemap and v-f-d internal.

e.g. 
In `https://unpkg.com/vue-facing-decorator@2.1.14/dist/esm/utils.js`, I import class `Base`(aka. `Vue`)  from './index', but in MyComp.js we import class `Vue` from `https://unpkg.com/vue-facing-decorator@2.1.14/dist/esm/index.js`. `index` and `index.js` are not the same file reference in browser, so we can't use `MyComp(which inherited from index.js) instanceof Vue(which imported from index)` to detect prototype chain.

unpkg redirect `index` to `index.js` automatically.

**Reply by johandanforth** [reply]
Alright I got things running now! I must say I like the style of v-f-d! Next thing is to get it working without referencing unpkg, and I'm not sure how that is going to work with the automatic redirection of xxx to xxx.js that unpkg does and that the internals in your code links to the base classes without the .js extension. I may have to create paths and aliases for all references in tsconfig.json, or what do you think?

If anyone else has the same problems I had, here's my test code:

cshtml/html
``` cshtml
<div id="app">
	<div class="row">
		<div class="col-6">
			<h1>My app</h1>
			<b>Count:</b> <span>{{ count }}</span><br />
			<button class="btn btn-primary" v-on:click="increment">increment</button>
			<fieldset>
				<legend>Messages:</legend>
				<pre>{{messages}}</pre>
			</fieldset>
		</div>
		<div class="col-6">
			<h3>My Component</h3>
			<my-component data-name="some name" v-on:name-changed="nameChanged" />
		</div>
	</div>

</div>

<!-- include the component template MyComponent.cshtml -->
<partial name="MyComponent" />

<script type="importmap">
	{
		"imports":{
			"vue": "https://unpkg.com/vue@3/dist/vue.esm-browser",
			"vue-facing-decorator":"https://unpkg.com/vue-facing-decorator@2.1.14/dist/esm/index",
			"MyApp":"./ts/MyApp.js"
		}
	}
</script>
<script type="module">
	import MyApp from 'MyApp'
	import { createApp } from 'vue'

	var app = createApp(MyApp);
	app.mount("#app");

</script>
```

MyApp.ts
``` typescript
import MyComponent from './MyComponent.js'

export default {
    data() {
        return {
            count: 0,
            messages: "mounted",
        }
    },
    mounted() {
        this.addMessage("MyApp mounted");
    },
    components: {
        MyComponent
    },
    methods: {
        increment() {
            this.count++;
            this.addMessage(`Count increased to ${this.count}`);
        },
        addMessage(message) {
            this.messages += `${message}\n`;
        },
        nameChanged(name) {
            this.addMessage(`Child component changed name to ${name}`);
        }
    }
}
```

MyComponent.cshtml (my vue component template)
``` vue
<template id="my-component">
	<div class="input-group">
		<div class="form-floating">
			<input v-model="name" class="form-control" placeholder="*" />
			<label>Change name to</label>
		</div>
		<button class="btn btn-primary" v-on:click="changeName">OK</button>
	</div>
</template>
```


MyComponent.ts 
``` typescript
import { Component, Emit, Prop, Vue } from 'vue-facing-decorator'

@Component({
    template: "#my-component"
})
export default class MyComponent extends Vue {

    @Prop
    dataName?: String;

    name = "name";

    mounted() {
        console.log("my-component mounted");
    }

    changeName() {
        this.nameChanged();
    }

    @Emit
    nameChanged() {
        return this.name;
    }

}
```

**Reply by ruojianll** [reply]
I suggest that remove `.js` in your code and do redirect like unpkg on your http server. That's compatible with ts style code and esm in broswer.

**Reply by johandanforth** [reply]
Yeah, I think I'll have to look at doing local redirects to the .js files in my server. Will have to look into that and get back. I wish there was another way of doing it though!

**Reply by johandanforth** [reply]
Yep, that worked perfectly well! I'm hosting my application in a dotnet core web application, so I put the libraries in a /lib/ folder and added a custom redirect in startup like this:

``` cs
app.Use(async (context, next) =>
{
	var url = context.Request.Path.Value;

	// Rewrite to .js if it's a lib request without .js extension
	if (url != null 
	    && url.ToLowerInvariant().StartsWith("/lib/") 
	    && !url.ToLowerInvariant().EndsWith(".js"))
	{
		context.Response.Redirect(url + ".js");
		return;  
	}

	await next();
});
```

And my importmapping looks like this now:

``` cshtml
<script type="importmap">
	{
		"imports":{
			"vue": "/lib/vue/dist/vue.esm-browser",
			"vue-facing-decorator":"/lib/vue-facing-decorator/dist/esm/index",
			"MyApp":"./ts/MyApp.js"
		}
	}
</script>
```

I still have to install vue and v-f-d as node modules for `tsc` to work properly. I should be able to get rid of that as well and point at the libraries in the /lib/ folder. Perhaps I can do that within `tsconfig.json` with path settings?

**Reply by ruojianll** [reply]
In my opinion, you must install v-f-d for `tsc`, the one wouldn't be compiled into output files, it just provides `.d.ts` files for type information.
